src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs

```diff
@@ -1124,13 +741,76 @@ protected internal unsafe bool WriteEvent(ref EventDescriptor eventDescriptor, I
+        /// <summary>
+        /// A struct characterizing ETW sessions (identified by the etwSessionId) as
+        /// activity-tracing-aware or legacy. A session that's activity-tracing-aware
+        /// has specified one non-zero bit in the reserved range 44-47 in the
+        /// 'allKeywords' value it passed in for a specific EventProvider.
+        /// </summary>
+        public struct SessionInfo
+        {
+            internal int sessionIdBit;      // the index of the bit used for tracing in the "reserved" field of AllKeywords
+            internal int etwSessionId;      // the machine-wide ETW session ID
+
+            internal SessionInfo(int sessionIdBit_, int etwSessionId_)
+            { sessionIdBit = sessionIdBit_; etwSessionId = etwSessionId_; }
+        }
+
+        private readonly WeakReference<EventProvider> m_eventProvider;
+        private long m_registrationHandle;
+        private GCHandle m_gcHandle;
+        private List<SessionInfo>? m_liveSessions;       // current live sessions (KeyValuePair<sessionIdBit, etwSessionId>)
+        private Guid m_providerId;
 
         internal EtwEventProvider(EventProvider eventProvider)
         {
-            _eventProvider = new WeakReference<EventProvider>(eventProvider);
+            m_eventProvider = new WeakReference<EventProvider>(eventProvider);
+        }
+
+        internal override void Disable()
+        {
+            base.Disable();
+            m_liveSessions = null;
+        }
+
+        protected override unsafe void HandleSessionEnable(
+                                    EventProvider target,
+                                    byte *additionalData,
+                                    byte level,
+                                    long matchAnyKeywords,
+                                    long matchAllKeywords,
+                                    Interop.Advapi32.EVENT_FILTER_DESCRIPTOR* filterData)
+        {
+            Debug.Assert(additionalData == null);
+
+            List<KeyValuePair<SessionInfo, bool>> sessionsChanged = GetSessions();
+
+            foreach (KeyValuePair<SessionInfo, bool> session in sessionsChanged)
+            {
+                int sessionChanged = session.Key.sessionIdBit;
+                int etwSessionId = session.Key.etwSessionId;
+                bool bEnabling = session.Value;
+
+                // reinitialize args for every session...
+                IDictionary<string, string?>? args = null;
+                ControllerCommand command = ControllerCommand.Update;
+
+                // if we get more than one session changed we have no way
+                // of knowing which one "filterData" belongs to
+                if (sessionsChanged.Count > 1)
+                    filterData = null;
+
+                // read filter data only when a session is being *added*
+                if (bEnabling)
+                {
+                    args = ParseFilterData(etwSessionId, filterData, out command);
```

Comment:
'Parse' strongly suggests that this call will solely be looking at the filterData that was passed in. Learning that Parse reads data from the registry was unexpected. 

How about factoring it something like this to make it more transparent what the code is doing
```C#
if (bEnabling)
{
    byte[] filterDataBytes;
    // if we get more than one session changed we have no way
    // of knowing which one "filterData" belongs to
    if (sessionsChanged.Count > 1 || filterData == null)
    {
        TryReadRegistryFilterData(etwSessionId, out filterDataBytes);
    }
    else
    {
        MarshalFilterData(filterData, out filterDataBytes, out command);
    }
    args = ParseFilterData(filterDataBytes);
}
```

The EventPipe version could just be MarshalFilterData(...) + ParseFilterData(...).